File: docs/sdk/python/pexipay/webhooks.py
"""Webhook signature verification"""

import hmac
import hashlib
import json
from typing import Any, Dict


def verify_webhook_signature(payload: str, signature: str, webhook_secret: str) -> bool:
    """
    Verify webhook signature from Pexipay

    Args:
        payload: Raw webhook payload (as string)
        signature: Signature from X-Pexipay-Signature header
        webhook_secret: Your webhook secret from Pexipay dashboard

    Returns:
        True if signature is valid
    """
    try:
        expected_signature = hmac.new(
            webhook_secret.encode("utf-8"), payload.encode("utf-8"), hashlib.sha256
        ).hexdigest()

        return hmac.compare_digest(signature, expected_signature)
    except Exception:
        return False


def construct_webhook_event(payload: str, signature: str, webhook_secret: str) -> Dict[str, Any]:
    """
    Parse and verify webhook event

    Args:
        payload: Raw webhook payload
        signature: Signature from header
        webhook_secret: Your webhook secret

    Returns:
        Parsed webhook event

    Raises:
        ValueError: If signature is invalid
    """
    is_valid = verify_webhook_signature(payload, signature, webhook_secret)

    if not is_valid:
        raise ValueError("Invalid webhook signature")

    return json.loads(payload)

File: docs/sdk/javascript/src/webhooks.ts
import * as crypto from 'crypto';

/**
 * Verify webhook signature from Pexipay
 * 
 * @param payload - Raw webhook payload (as string)
 * @param signature - Signature from X-Pexipay-Signature header
 * @param webhookSecret - Your webhook secret from Pexipay dashboard
 * @returns True if signature is valid
 */
export function verifyWebhookSignature(
  payload: string | Buffer,
  signature: string,
  webhookSecret: string
): boolean {
  try {
    const payloadString = typeof payload === 'string' ? payload : payload.toString('utf8');
    
    const expectedSignature = crypto
      .createHmac('sha256', webhookSecret)
      .update(payloadString)
      .digest('hex');

    return crypto.timingSafeEqual(
      Buffer.from(signature),
      Buffer.from(expectedSignature)
    );
  } catch (error) {
    return false;
  }
}

/**
 * Parse and verify webhook event
 * 
 * @param payload - Raw webhook payload
 * @param signature - Signature from header
 * @param webhookSecret - Your webhook secret
 * @returns Parsed webhook event
 * @throws Error if signature is invalid
 */
export function constructWebhookEvent<T = any>(
  payload: string | Buffer,
  signature: string,
  webhookSecret: string
): T {
  const isValid = verifyWebhookSignature(payload, signature, webhookSecret);
  
  if (!isValid) {
    throw new Error('Invalid webhook signature');
  }

  const payloadString = typeof payload === 'string' ? payload : payload.toString('utf8');
  return JSON.parse(payloadString);
}
